[MRG+2] #7908 Addressed issue in first iteration of RANSAC regression

mthorrell opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
Fixes #7908 


#### What does this implement/fix? Explain your changes.
On the first iteration of RANSAC regression, if no inliers are found, an error is produced and the code is stopped.  Ideally the procedure would just skip that iteration and continue on to the next iteration where it would use a different random sample which could produce valid inliers.

Generally this error is produced when `n_inliers_subset` and `n_inliers_best` are both zero. My fix was to set the initial value for `n_inliers_best` to 1.  Thus if `n_inliers_subset >= 1`, the code follows the normal path, and if `n_inliers_subset == 0`, the code progresses to the next iteration. This fixes the bug as described in this issue.

However, setting `n_inliers_best = 1` creates an issue again in the first iteration of this loop in the case when `n_inliers_subset == 1`. The subsequent comparison is made: `score_subset < score_best`. Since `score_best` is initialized to `np.inf`, the code will incorrectly skip to the next iteration, ignoring the fact that in the first iteration, we did find valid inliers.  The fix for this is to change the initial value of `score_best` to `-np.inf`.  In general, I think this is better practice when initializing these types of variables anyway.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
Addressed issue 7908 in RANSAC regression

Commit:
changed np.inf to -np.inf for first iteration

Commit:
Changed assert_raises_regexp to assert_raises

mthorrell commented:
In this small update, I adjusted the offending test so it did not require a specific output; however, an error is still required to be raised in the residual threshold = 0 case.

amueller commented:
thanks for the PR. I think changing the tests was the way to go! (I don't have time for a review right now unfortunately)

Commit:
included error message in test. cleaned up error message format

mthorrell commented:
Thanks. 

I made a couple minor changes:
- The test now more specifically checks the error message since I think this is better practice.
- I also cleaned up the formatting of that error message.

Commit:
added whitespaces after comma

amueller commented on sklearn/linear_model/tests/test_ransac.py:
```diff
@@ -153,9 +153,14 @@ def test_ransac_resid_thresh_no_inliers():
+    msg = (
+        "RANSAC could not find valid consensus set, because"
+        " either the `residual_threshold` rejected all the samples or"
+        " `is_data_valid` and `is_model_valid` returned False for all"
```
refering to internal functions is not very helpful for the users imho.

mthorrell replied:
I believe you are referring to `is_data_valid` and `is_model_valid`. These are user-defined functions much like `residual_threshold` is a user-defined parameter.  When a user does not define these functions, RANSAC does not call them--they are initialized to `None` and evaluation is skipped. Unless you disagree, these may still be valid to refer to.

Perhaps a better message might read:
 
RANSAC could not find a valid consensus set. All `max_trials` iterations were skipped because each randomly chosen sub-sample either: didn't produce a valid set of inliers due to a small `residual_threshold` or was invalidated by `is_data_valid` or `is_model_valid` if either of these functions have been defined by the user.

amueller replied:
Oh, sorry, my fault. The current message is ok, though your suggestion is even better!

amueller commented:
maybe @ahojnnes wants to review?

Commit:
changed stop message

jnothman commented on sklearn/linear_model/tests/test_ransac.py:
```diff
@@ -153,9 +153,15 @@ def test_ransac_resid_thresh_no_inliers():
+        " iterations were skipped because each randomly chosen"
+        " sub-sample either: didn't produce a valid set of inliers"
+        " due to a small `residual_threshold` or was invalidated by"
+        " `is_data_valid` or `is_model_valid` if either of these"
```
I think the "if either ..." can be removed

jnothman reviewed (commented):
I am not certain this is the right fix. This still fails if a 0-inlier sample is drawn at any point after the first >0-inlier sample is drawn.

We could introduce a parameter to control how many 0-inlier samples are allowed through, or just get rid of this control (the main risk being lots of time wasted).

A couple of more cosmetic things we can do to improve this behaviour:
* If we continue to raise an error when no inliers are found, we should report the minimum residual found (and perhaps other summary statistics of the residual distribution) as well as the current threshold, so that the user has a means to tune the parameter.
* introduce a verbose option that reports things like the median residual, number of inliers, etc., at each iteration.

Ping @CSchoel for your opinion.

jnothman commented:
Sorry, silly me. This patch does indeed fix the issue because the check for `n_inliers_subset == 0` only applies when it is >= the best subset (defaulting to 1 in this PR). But that seems to mean that the "No inliers found" error is unreachable; and as a result we may draw all trials with 0 samples and this is expensive.

*If* this is the right patch, you need to write a test. But we may want a more configurable parameter to control the number of 0-inlier samples. Certainly, we should be giving the user more information on how to set the threshold appropriately.

mthorrell commented:
The `if` statement containing `n_inliers_subset == 0` was unreachable and hence was removed, so there shouldn't be any unreachable code generated by this PR.

If this is the fix, a test already exists that generates an appropriate error. Unless I'm missing something, there would be nothing else to do here except potentially edit the error message--again assuming we go with this fix.

However, I do think it makes sense to put a limit on the number of iterations that are tried which end in skips. Also, `is_data_valid` or `is_model_valid` themselves could be computationally expensive, so I would think we would count skips due to data/model failure the same as skips due to a strict `residual_threshold`. I suppose one could explicitly track the causes of different skips separately (`is_data_valid` vs. `is_model_valid` vs. `residual_threshold`), but this seems excessive to me.

I'll submit a commit that allows a user to set a `max_skips` parameter. It will default to `max_trials`.

jnothman commented:
(Sorry for my confusions. Was trying to grok the overall algorithm)

jnothman commented:
> If this is the fix, a test already exists that generates an appropriate error. 

A non-regression test, which allows 0 inliers to be drawn in the first iteration, would be ideal.

jnothman commented:
Yes, I don't mind `max_skips`. I also don't think there's any harm in reporting diagnostics of various kinds, summarising residuals or reason for skipping / not updating.

Commit:
tracked skipped iterations added errors and warnings

Commit:
fixed documentation

mthorrell commented:
err... need to finish the tests, so this is still WIP

Commit:
added tests and some further cleanup

Commit:
removed unnecessary import

jnothman commented:
Still still WIP?

jnothman commented on sklearn/linear_model/ransac.py:
```diff
@@ -111,6 +111,11 @@ class RANSACRegressor(BaseEstimator, MetaEstimatorMixin, RegressorMixin):
 
+    max_skips : int, optional
+        Maximum number of iterations that can be skipped due to finding zero
+        inliers or invalid data defined by `is_data_valid` or invalid models
```
need double-backticks

jnothman reviewed (commented):
It would be useful to set these diagnostics as attributes of the model, i.e. `n_skips_no_inliers_`, etc.

We may also then choose to remove them from the error message (and certainly from the warning, to help the warnings module ignore duplicates), pointing the user to these attributes instead.

jnothman reviewed (commented):
I'd still consider including a diagnostic about the stringency of the threshold, either reporting some statistic of the residuals (e.g. average min) or some statistic of the inlier subset size.

mthorrell commented:
I like the attributes idea. The error and warning messages seemed a bit clunky.  Let me see if I can do something about the stringency of the threshold as well.

Commit:
added skip count attributes and adjusted tests

Commit:
made changes to other error messages

mthorrell commented:
I added the `n_skips_no_inliers` etc. attributes and adjusted the error messages and tests.  

I did not add anything indicating the stringency of the threshold.  Should that be different PR?  This one is getting rather large imo.

jnothman commented:
Okay. Large? Not so much. Unnecessarily long in the tooth? I agree. Let's
leave scope as it is currently.

On 29 December 2016 at 05:20, mthorrell <notifications@github.com> wrote:

> I added the n_skips_no_inliers etc. attributes and adjusted the error
> messages and tests.
>
> I did not add anything indicating the stringency of the threshold. Should
> that be different PR? This one is getting rather large imo.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7914#issuecomment-269517499>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6w1udNqqptWBXiWpFDhR-CxM156fks5rMqhSgaJpZM4K3b8L>
> .
>

jnothman commented on sklearn/linear_model/ransac.py:
```diff
@@ -111,6 +111,11 @@ class RANSACRegressor(BaseEstimator, MetaEstimatorMixin, RegressorMixin):
+    max_skips : int, optional
+        Maximum number of iterations that can be skipped due to finding zero
+        inliers or invalid data defined by ``is_data_valid`` or invalid models
+        defined by ``is_data_valid``.
```
should be is_model_valid

jnothman commented on sklearn/linear_model/ransac.py:
```diff
@@ -395,12 +418,36 @@ def fit(self, X, y, sample_weight=None):
+                    "RANSAC skipped more iterations than `max_skips` without"
+                    " finding a valid consensus set. Iterations were skipped"
+                    " because each randomly chosen sub-sample failed the"
+                    " passing criteria. The object attributes"
```
object -> estimator. But perhaps "See estimator attributes for diagnostics (n_skips*)." is sufficient

jnothman reviewed (approved):
Otherwise LGTM!

Commit:
fixed attribute desc. and changed error/warn messages

agramfort commented:
+1 for merge

@jnothman merge if you're happy

jnothman commented on sklearn/linear_model/ransac.py:
```diff
@@ -111,6 +111,11 @@ class RANSACRegressor(BaseEstimator, MetaEstimatorMixin, RegressorMixin):
+        Maximum number of iterations that can be skipped due to finding zero
+        inliers or invalid data defined by ``is_data_valid`` or invalid models
+        defined by ``is_model_valid``.
+
```
I suppose nowadays we're meant to have `.. versionadded` annotations in docstrings.

mthorrell replied:
Is it accurate to say `.. versionadded:: 0.19` or should it still be `0.18`?

jnothman commented:
0.19.

On 3 January 2017 at 05:14, mthorrell <notifications@github.com> wrote:

> *@mthorrell* commented on this pull request.
> ------------------------------
>
> In sklearn/linear_model/ransac.py
> <https://github.com/scikit-learn/scikit-learn/pull/7914>:
>
> > @@ -111,6 +111,11 @@ class RANSACRegressor(BaseEstimator, MetaEstimatorMixin, RegressorMixin):
>      max_trials : int, optional
>          Maximum number of iterations for random sample selection.
>
> +    max_skips : int, optional
> +        Maximum number of iterations that can be skipped due to finding zero
> +        inliers or invalid data defined by ``is_data_valid`` or invalid models
> +        defined by ``is_model_valid``.
> +
>
> Is it accurate to say .. versionadded:: 0.19 or should it still be 0.18?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7914>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAEz69Y4ouiibM8QLXSVzox9SqGhbbY-ks5rOT6jgaJpZM4K3b8L>
> .
>

Commit:
added versionadded

mthorrell commented:
made the `versionadded` change. I believe this is finished, assuming the tests pass.

jnothman reviewed (commented):
Sorry, I'd forgotten to ask you to add an entry to `whats_new.rst` (under enhancements I think? bug fixes if you rather?)

Commit:
changed the whats new doc

mthorrell commented:
I made the entry in `whats_new.rst`. Is there anything else needed?

jnothman committed:
Merge branch 'master' into ransac_bugfix

jnothman commented:
All good, thanks!
